File: tests/interop/python-client/interop_client.py
#!/usr/bin/env python3
"""Interop test client -- Python

Connects to the KCP MCP bridge via stdio, lists resources and reads each one,
then prints structured results as JSON for parity comparison.

Usage: python interop_client.py <manifest-path> [output-path]
"""

import asyncio
import json
import sys
from pathlib import Path

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


async def main():
    manifest_path = sys.argv[1] if len(sys.argv) > 1 else None
    output_path = sys.argv[2] if len(sys.argv) > 2 else "python-results.json"

    if not manifest_path:
        print("Usage: python interop_client.py <manifest-path> [output-path]",
              file=sys.stderr)
        sys.exit(1)

    manifest_path = str(Path(manifest_path).resolve())

    # Find the Python bridge
    repo_root = Path(__file__).resolve().parent.parent.parent.parent
    bridge_python = repo_root / "bridge" / "python"

    # Use the Python bridge via stdio
    server_params = StdioServerParameters(
        command=sys.executable,
        args=["-m", "kcp_mcp", manifest_path],
        cwd=str(bridge_python),
    )

    resources = []

    async with stdio_client(server_params) as (read_stream, write_stream):
        async with ClientSession(read_stream, write_stream) as session:
            await session.initialize()

            # List all resources
            list_result = await session.list_resources()

            for res in list_result.resources:
                entry = {
                    "uri": str(res.uri),
                    "name": res.name,
                    "description": getattr(res, "description", None),
                    "mimeType": getattr(res, "mimeType", None),
                }

                # Read each resource
                try:
                    read_result = await session.read_resource(res.uri)
                    contents = read_result.contents
                    if contents and len(contents) > 0:
                        first = contents[0]
                        text = getattr(first, "text", None)
                        if text:
                            entry["content_preview"] = text[:200]
                except Exception as e:
                    print(f"Failed to read {res.uri}: {e}", file=sys.stderr)

                resources.append(entry)

    # Sort by URI for deterministic comparison
    resources.sort(key=lambda r: r["uri"])

    result = {
        "resource_count": len(resources),
        "resources": resources,
    }

    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(result, f, indent=2)

    print(f"Wrote {len(resources)} resources to {output_path}")


if __name__ == "__main__":
    asyncio.run(main())

File: tests/interop/ts-client/interop_client.ts
/**
 * Interop test client -- TypeScript
 *
 * Connects to a KCP MCP bridge via stdio, lists resources and reads each one,
 * then prints structured results as JSON for parity comparison.
 *
 * Usage: npx tsx interop_client.ts <bridge-command> <bridge-args...> -- <output-path>
 *   or:  npx tsx interop_client.ts <manifest-path> [output-path]
 */

import { resolve, dirname } from "node:path";
import { writeFileSync } from "node:fs";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";

interface ResourceEntry {
  uri: string;
  name: string;
  description?: string;
  mimeType?: string;
  content_preview?: string;
}

interface InteropResult {
  resource_count: number;
  resources: ResourceEntry[];
}

async function main(): Promise<void> {
  const manifestPath = process.argv[2];
  const outputPath = process.argv[3] || "ts-results.json";

  if (!manifestPath) {
    console.error("Usage: npx tsx interop_client.ts <manifest-path> [output-path]");
    process.exit(1);
  }

  const absManifest = resolve(manifestPath);

  // Start the TS bridge as a subprocess with stdio transport
  const bridgePath = resolve(
    dirname(new URL(import.meta.url).pathname),
    "../../../bridge/typescript/dist/cli.js"
  );

  const transport = new StdioClientTransport({
    command: "node",
    args: [bridgePath, absManifest],
  });

  const client = new Client({
    name: "interop-ts-client",
    version: "1.0.0",
  });

  await client.connect(transport);

  // List all resources
  const listResult = await client.listResources();
  const resources: ResourceEntry[] = [];

  for (const res of listResult.resources) {
    const entry: ResourceEntry = {
      uri: res.uri,
      name: res.name,
      description: res.description,
      mimeType: res.mimeType,
    };

    // Read each resource to get content
    try {
      const readResult = await client.readResource({ uri: res.uri });
      const contents = readResult.contents;
      if (contents && contents.length > 0) {
        const firstContent = contents[0];
        if ("text" in firstContent && typeof firstContent.text === "string") {
          // Store first 200 chars as preview for comparison
          entry.content_preview = firstContent.text.substring(0, 200);
        }
      }
    } catch (err: any) {
      console.error(`Failed to read ${res.uri}: ${err.message}`);
    }

    resources.push(entry);
  }

  // Sort by URI for deterministic comparison
  resources.sort((a, b) => a.uri.localeCompare(b.uri));

  const result: InteropResult = {
    resource_count: resources.length,
    resources,
  };

  writeFileSync(outputPath, JSON.stringify(result, null, 2));
  console.log(`Wrote ${resources.length} resources to ${outputPath}`);

  await client.close();
  process.exit(0);
}

main().catch((err) => {
  console.error(`Fatal: ${err.message}`);
  process.exit(1);
});
